add cdi-alternative quickstart

nzhu972 opened the pull request:
This is a Quick-Start showing how CDI-Alternative is implemented. The project consists of an EAR with a nested JAR (ejb) and WAR.

Commit:
add cdi-alternative quickstart

sgilda commented on cdi-alternative/README.md:
```diff
@@ -0,0 +1,88 @@
+CDI_Alternative: Demostration of CDI Alternative implementation
```
This should match the folder and quickstart name. In other words:
cdi-alternative: Demonstrates CDI Alternatives

sgilda commented on cdi-alternative/README.md:
```diff
@@ -0,0 +1,88 @@
+CDI_Alternative: Demostration of CDI Alternative implementation
+======================================================
+Author: Nevin Zhu
+
```
You are missing some metadata that is used to build the quickstart table in the root README.html file and the jdf-site. 

Level: Intermediate
Technologies: CDI, 
Summary: Demonstrates the use of CDI to provide the alternative choices at deployment time. (you might be able to word this better)
Target Product: EAP

For the technologies, you can see what the other quickstarts list here:  (not sure what else it contains, so feel free to look at the others here: http://www.jboss.org/jdf/quickstarts/get-started/)

sgilda commented on cdi-alternative/README.md:
```diff
@@ -0,0 +1,88 @@
+
+    *  To handle client-specific business logic that is determined at runtime
+    *  To specify beans that are valid for a particular deployment scenario (for example, when country-specific sales tax laws require country-specific sales tax business logic)
+    *  To create dummy (mock) versions of beans to be used for testing
```
The above 3 lines should not be indented. They appear strangely when rendered in html. :-)

sgilda commented:
Nevin, I am not able to build and deploy your quickstart. It looks like you're using the EAP 6 version of the artifacts in your pom.xml file. These are community quickstarts so we need to use the community artifacts. 

Also, we no longer put versions throughout the files. If you look at the other quickstarts in this repository, you'll see that they all now define properties for the versions, then they use those properties throughout the file. 

For example, if you look at the greeter/pom.xml file, you'll see:

```
    <properties>
        <!-- Explicitly declaring the source encoding eliminates the following message: -->
        <!-- [WARNING] Using platform encoding (UTF-8 actually) to copy filtered 
                 resources, i.e. build is platform dependent! -->
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>

        <!-- JBoss dependency versions -->
        <version.org.jboss.as.plugins.maven.plugin>7.3.Final</version.org.jboss.as.plugins.maven.plugin>
        <version.org.jboss.spec.jboss.javaee.6.0>3.0.0.Final</version.org.jboss.spec.jboss.javaee.6.0>

        <!-- other plugin versions -->
        <version.compiler.plugin>2.3.1</version.compiler.plugin>
        <version.war.plugin>2.1.1</version.war.plugin>

        <!-- maven-compiler-plugin -->
        <maven.compiler.target>1.6</maven.compiler.target>
        <maven.compiler.source>1.6</maven.compiler.source>
    </properties>
```

Then, later in that file, it uses the property like this:

```
    <dependency>
        <groupId>org.jboss.spec</groupId>
        <artifactId>jboss-javaee-6.0</artifactId>
        <version>${version.org.jboss.spec.jboss.javaee.6.0}</version>
        <type>pom</type>
        <scope>import</scope>
    </dependency>
```

You will need to modify your pom.xml file to use properties and the community versions of of the artifacts.

sgilda commented:
Since this is an EAR, you might want to check out the other quickstarts that build EARs. The properties are defined in the top level pom.xml file but used in the subprojects. Examples are kitchensink-ear and ejb-in-ear.

pmuir commented:
Sande,

Let me know when you want me to review this qs.

Pete

sgilda commented:
@pmuir : If you want, you can do a code review now. The pom.xml file needs to be fixed before we can build and test it though.

Commit:
Updated Readme and properties in POM

pmuir commented:
Wouldn't this be simpler as a war, we can then remove about half the files ;-)

pmuir commented on cdi-alternative/ejb/src/main/java/org/jboss/as/quickstarts/cdi/alternative/Tax.java:
```diff
@@ -0,0 +1,32 @@
+ * 
+ */
+@Remote
+public interface Tax {
```
Having all these Tax classes is quite confusing, might want to give the Tax ones a separate name?

pmuir commented on cdi-alternative/ejb/src/main/java/org/jboss/as/quickstarts/cdi/alternative/TaxBean.java:
```diff
@@ -0,0 +1,44 @@
+ * @author Nevin Zhu
+ * 
+ */
+public @Singleton class TaxBean implements Tax {
```
We normally put annotations before keywords.

pmuir commented on cdi-alternative/web/src/main/java/org/jboss/as/quickstarts/cdi/alternative/web/Demo.java:
```diff
@@ -0,0 +1,74 @@
+	 */
+	protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+		// TODO Auto-generated method stub
+	}
```
Fix this comment.

pmuir commented on cdi-alternative/README.md:
```diff
@@ -0,0 +1,88 @@
+
+The application will be running at the following URL <http://localhost:8080/cdi-alternative-web/Demo>.
+
+A message will display for the bean being injected during run time. The alternative version of the beans can be updated in beans.xml
```
It would be nice to walk people through changing the alternative in use.

sgilda replied:
Agreed!

sgilda commented:
Nevin, your email said you had made the fixes, yet I don't see an update in your repository. Did you remember to push your local file changes to the branch on your repository?

Commit:
simplied project by removing ejb and ear. War only.

nzhu972 commented:
@pmuir 
I took your input and revised the project. It is now a WAR deployment only. Stripped out the EJB project and updated the pom.xml accordingly. I renamed the tax class file names as well to something more generic. Let me know how it looks. The changes are committed and pushed.

sgilda commented:
Hi Nevin,
I'm a little confuse by your new directory structure. This is what I see:

./ear/pom.xml
./ejb/src/main/java/org/jboss/as/quickstarts/cdi/alternative/California.java
./ejb/src/main/java/org/jboss/as/quickstarts/cdi/alternative/StateTax.java
./ejb/src/main/java/org/jboss/as/quickstarts/cdi/alternative/Tax.java
./ejb/src/main/java/org/jboss/as/quickstarts/cdi/alternative/TaxBean.java
./ejb/src/main/java/org/jboss/as/quickstarts/cdi/alternative/Washington.java.
./ejb/src/main/resources/META-INF/beans.xml
./web/src/main/webapp/result.jsp
./web/src/main/java/org/jboss/as/quickstarts/cdi/alternative/web/Demo.java
./src/main/webapp/result.jsp
./src/main/webapp/WEB-INF/beans.xml
./src/main/java/org/jboss/as/quickstarts/cdi/alternative/Demo.java
./src/main/java/org/jboss/as/quickstarts/cdi/alternative/Tax.java
./src/main/java/org/jboss/as/quickstarts/cdi/alternative/TaxImpl_1.java
./src/main/java/org/jboss/as/quickstarts/cdi/alternative/TaxImpl_2.java

You have an ear/ directory that only contains a pom.xml file.
You have duplicate class names in the ejb/ and src/ directory structures.
You have 2 beans.xml files: one in the ejb/ directory and one in the src/ directory
You have 2 result.jsp files, one in the ejb/ directory structure and one in teh src/ directory structure.

Should the `ear` and `ejb` directories still be there?

nzhu972 commented:
Hi Sande,

ear, ejb and web directories are all removed. There should just be a src directory along with one pom.xml. 
I executed a ' git add cdi-alternative/\* ' and i expect the command to remove the directories that no longer exist.I will look at the git commands. Perhaps I skipped a step. 

-Nevin

sgilda commented:
Hi Nevin,

You can see the files included in this pull by clickin on the files tab, https://github.com/jboss-jdf/jboss-as-quickstart/pull/407/files, or by looking at the commits. There you will see the list I displayed above. 

If those files were included in a previous commit, you will need to do a 'git rm' to remove them.  At some point you should probably squash your commits down to just 1 commit too to make it easier to review. :-)

Sande

pmuir commented:
Nevin, you'll need to `git rm` the files that you want to remove from git...

nzhu972 commented:
Ran into a new problem when committing today. It kept complaining that my branch needs a merge but even after executing a git pull, it wasn't going anywhere. I dont want to continue spending time troubleshooting GIT, it's eating wayy to much time into it so I blew my existing clone away and started over. I plan to create a new topic branch with the new code then open a new pull request. Is that a sound approach? There should be a way to blow away the existing branch that I first created. correct?

LightGuard commented:
On Tue, Jan 29, 2013 at 10:22 AM, nzhu972 notifications@github.com wrote:

> Ran into a new problem when committing today. It kept complaining that my
> branch needs a merge but even after executing a git pull, it wasn't going
> anywhere. I dont want to continue spending time troubleshooting GIT, it's
> eating wayy to much time into it so I blew my existing clone away and
> started over. I plan to create a new topic branch with the new code then
> open a new pull request. Is that a sound approach? There should be a way to
> blow away the existing branch that I first created. correct?
> 
> git branch -D <branch_name>

The other way will work also, just depends how much you want to dive into
Git right now.

>  —
> Reply to this email directly or view it on GitHubhttps://github.com/jboss-jdf/jboss-as-quickstart/pull/407#issuecomment-12846176.

## 

Jason Porter
http://en.gravatar.com/lightguardjp

nzhu972 commented:
Ok. I have opened a new branch 'mycontribution' and created a separate pull request. Can someone close this pull request. thanks.

sgilda commented:
Closed this pull since it's been reissued under another pull.

sgilda commented:
Nevin,
Here are some tips for the future:

If you have files that were committed and should not have been, you can delete them from your local file system, then when you do a 'git status', Git will give you a list of the files that were removed. I usually make a script from that output to issue a 'git rm <filename>' for each file name that I deleted. 

After that, do a git status and if everything looks good, you can do a 'git push -f <your-github> HEAD to push the changes to your Git. The '-f' forces the push. I renamed my own github remote from 'origin' to 'sgilda' to make it clearer, so I do 'git push -f sgilda HEAD'. Then I look at the branch on my Git to verify that what I pushed looks valid. 

Hope this helps.
